include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -1068,50 +1092,349 @@ namespace Differentiation
+       * The internal number type code.
+       */
+      static constexpr enum NumberTypes type_code = NumberTypes::none;
+
+      // The clang compiler does not seem to like these
+      // variables being defined as constant expressions
+      // (the tests <adolc|sacado>/ad_number_traits_02 will
+      // fail with linking errors). However, GCC complains
+      // about the use of non-constant expressions in
+      // std::conditional.
+#  ifdef __clang__
+
+      /**
+       * A flag to indicate whether the number is of
+       * the taped variety or not.
+       */
+      static const bool is_taped;
+
+
+      /**
+       * A flag to indicate whether the number is of
+       * the tapeless variety or not.
+       */
+      static const bool is_tapeless;
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a real value.
+       */
+      static const bool is_real_valued;
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a complex value.
+       */
+      static const bool is_complex_valued;
+
+
+      /**
+       * The number of directional derivatives that can be
+       * taken with this (floating point) number.
+       */
+      static const unsigned int n_supported_derivative_levels;
+
+#  else
+
+      /**
+       * A flag to indicate whether the number is of
+       * the taped variety or not.
+       */
+      static constexpr bool is_taped = false;
+
+
+      /**
+       * A flag to indicate whether the number is of
+       * the tapeless variety or not.
+       */
+      static constexpr bool is_tapeless = false;
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a real value.
+       */
+      static constexpr bool is_real_valued =
+        (!boost::is_complex<ScalarType>::value);
+
+
+      /**
+       * A flag to indicate whether the number represents
+       * a complex value.
+       */
+      static constexpr bool is_complex_valued = !is_real_valued;
+
+
+      /**
+       * The number of directional derivatives that can be
+       * taken with this (floating point) number.
+       */
+      static constexpr unsigned int n_supported_derivative_levels = 0;
+
+#  endif
+
+
       /**
        * Underlying floating point value type.
        * This could real-valued or complex-valued.
        */
       using scalar_type = ScalarType;
 
-      static ScalarType
-      get_directional_derivative(const ScalarType & /*x*/,
-                                 const unsigned int /*direction*/)
+
+      /**
+       * Type for real numbers.
+       */
+      using real_type =
+        typename dealii::numbers::NumberTraits<scalar_type>::real_type;
+
+
+      /**
+       * Type for complex numbers.
+       */
+      using complex_type = std::complex<real_type>;
+
+
+      /**
+       * The actual auto-differentiable number type.
+       */
+      using ad_type = typename std::
+        conditional<is_real_valued, real_type, complex_type>::type;
```

Comment:
Is this (and the `derivative_type` below) not equivalent to `scalar_type` (or `ScalarType`)?

Reply:
Probably... :-D